modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
+              result.buff[i] = b - 0x20;
+            }
+          } else { // last == LETTER
+            if _byte_isUpper(b) {
+              result.buff[i] = b + 0x20;
+            }
+          }
+        } else {
+          // Uncased elements
+          last = UN;
+        }
+      }
+      return result;
+    }
+
+    proc capitalize() /*: string*/ {
+      var result: string = this.toLower();
+      if result.isEmptyString() then return result;
+
+      var b = result.buff[0];
+      if _byte_isLower(b) { // Only change alpha
+        result.buff[0] = b - 0x20;
+      }
+      return result;
+    }
+
+  } // end record string
+
+
+  // We'd like this to be by ref, but doing so leads to an internal
+  // compiler error.  See
+  // $CHPL_HOME/test/types/records/sungeun/recordWithRefCopyFns.future
+  pragma "donor fn"
+  pragma "auto copy fn"
+  proc chpl__autoCopy(s: string) {
+    if debugStrings then
+      chpl_debug_string_print("in autoCopy()");
+
+    // This pragma may be unnecessary.
+    pragma "no auto destroy"
+    var ret: string;
+    const slen = s.len; // cache the remote copy of len
+    if slen != 0 {
+      if s.locale.id == chpl_nodeID {
+        if debugStrings then
+          chpl_debug_string_print("  local initCopy");
+        ret.buff = chpl_mem_alloc(s._size.safeCast(size_t),
+                                  CHPL_RT_MD_STR_COPY_DATA): bufferType;
+        memmove(ret.buff, s.buff, s.len.safeCast(size_t));
+        ret.buff[s.len] = 0;
+      } else {
+        if debugStrings then
+          chpl_debug_string_print("  remote initCopy: "+s.locale.id:c_string);
+        ret.buff = copyRemoteBuffer(s.locale.id, s.buff, slen);
+      }
+      ret.len = slen;
+      ret._size = slen+1;
+      ret.owned = true;
+    }
+
+    if debugStrings then
+      chpl_debug_string_print("leaving autoCopy()");
     return ret;
   }
-  
-  inline proc _string_contains(a: string, b: string)
-    return _string_contains(a.c_str(), b.c_str());
 
-  /* args: any number of strings
-     return: Returns true if this starts with one of the strings specified in args
+  /*
+   * NOTES: Just a word on memory allocation in the generated code.
+   * Any function that returns a string copies the returned value
+   * including the c_string via a call to chpl__initCopy().  The
+   * copied value is returned, and the original is destroyed on the
+   * way out.  I'd really like to figure out a way to pass back the
+   * original string without copying it.  I think I might be able to
+   * do it by putting some sort of flag in the string record that is
+   * used by initCopy().
+   * TODO: Check if ^ is still true w/ the new AMM
+   * TODO: Do we need an initCopy for strings?  If not, this clause can be removed.
    */
-  inline proc string.startsWith(args ...?k):bool {
-    for param i in 1..k {
-      if (this.substring(0..args(i).length) == args(i))
-        then return true;
+  pragma "init copy fn"
+  proc chpl__initCopy(s: string) {
```

Comment:
I don't understand why we have autoCopy and initCopy and how they are different. I don't think we plan to have both in the future. But, my notes to autoCopy apply here too.